snoop/data/management/commands/filestats.py

```diff
@@ -0,0 +1,128 @@
+    'application/vnd.openxmlformats-officedocument.wordprocessingml.template',
+    'application/vnd.ms-word.document.macroEnabled.12',
+    'application/vnd.ms-word.template.macroEnabled.12',
+    'application/vnd.oasis.opendocument.text',
+    'application/vnd.oasis.opendocument.text-template',
+    'application/rtf',
+
+    'application/vnd.ms-excel',
+    'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
+    'application/vnd.openxmlformats-officedocument.spreadsheetml.template',
+    'application/vnd.ms-excel.sheet.macroEnabled.12',
+    'application/vnd.ms-excel.template.macroEnabled.12',
+    'application/vnd.ms-excel.addin.macroEnabled.12',
+    'application/vnd.ms-excel.sheet.binary.macroEnabled.12',
+    'application/vnd.oasis.opendocument.spreadsheet-template',
+    'application/vnd.oasis.opendocument.spreadsheet',
+
+    'application/vnd.openxmlformats-officedocument.presentationml.presentation',
+    'application/vnd.openxmlformats-officedocument.presentationml.template',
+    'application/vnd.openxmlformats-officedocument.presentationml.slideshow',
+    'application/vnd.ms-powerpoint',
+    'application/vnd.ms-powerpoint.addin.macroEnabled.12',
+    'application/vnd.ms-powerpoint.presentation.macroEnabled.12',
+    'application/vnd.ms-powerpoint.template.macroEnabled.12',
+    'application/vnd.ms-powerpoint.slideshow.macroEnabled.12',
+    'application/vnd.oasis.opendocument.presentation',
+    'application/vnd.oasis.opendocument.presentation-template',
+
+    'application/xhtml+xml',
+    'application/xml',
+    'text/xml',
+    'image/tiff',
+    'image/jpg',
+    'image/webp',
+    'image/heic',
+
+    'application/vnd.ms-outlook',
+    'application/vnd.ms-office',
+    'application/CDFV2',
+    'message/rfc822',
+    'message/x-emlx',
+}
+
+def truncateSize(size):
+    exponent = len(str(size))-1
+    divisor = 10**exponent
+    return int((size/divisor))*divisor
+
+class Command(BaseCommand):
+    help = "Display filetype stats."
+
+    def add_arguments(self, parser):
+        parser.add_argument('collection', type=str)
+
+    def handle(self, collection, **options):
+
+        col = collections.ALL[collection]
+        
+        with col.set_current():
+            querysetMime = models.Blob.objects.all().values('mime_type').annotate(total=Count('mime_type')).annotate(size=Sum('size')).order_by('-size')
+            print()
+            print('---- Top 100 Mime-Types by size ----')
+            print()
+            for mtype in querysetMime:
+                print(f'{(mtype["mime_type"]):75} {truncateSize(mtype["size"]):10d}')
+
+            print()
+            print('---- Top Unsupported Mime-Types ----')
+            print()
+            unsuppQuery = querysetMime.exclude(mime_type__in=SUPPORTED_FILETYPES)
+            for mtype in unsuppQuery[:100]:
+                #if mtype['mime_type'] not in SUPPORTED_FILETYPES:
+                extensions = mimetypes.guess_all_extensions(mtype['mime_type'])
```

Comment:
here I want to get the sizes for the actual extensions from the actual files on disk from the `blob.file_set`

the filename field you have to split is binary, see here: https://github.com/liquidinvestigations/hoover-snoop2/blob/10941701ea6bda5a8efe27bcceb48cce19616a4f/snoop/data/models.py#L186

you probably want to use these nasty functions here https://docs.djangoproject.com/en/3.1/ref/models/database-functions/#django.db.models.functions.StrIndex that translate to their nasty SQL counterparts for doing this; i'm sure they work just fine on binary fields too (just be sure to give them `b'...'` strings)

please keep this if possible to one single query (like the ones for mime-type) since looping in python between 2 different queries will take forever on a 2TB dataset. also please display the extensions one per line, sorted after size, while also displaying the mime-type for each extension

Reply:
or just do this

```
size_for_ext = defaultdict(int)
mime_for_ext = defaultdict(set)
for f in File.objects.all().iterator():
    ext = f.name_bytes.split(b'.')[-1][:100]
    size_for_ext[ext] += f.size
    mime_for_ext[ext].add(f.mime_type)

print(size_for_ext)
print(mime_for_ext)